### PR TITLE
Add padding-free training for vision datasets in SFTTrainer

### LINKED ISSUE #4339: Packing with VLMs
### Feature request

Per https://github.com/axolotl-ai-cloud/axolotl/issues/3131, would it be possible to implement packing (in SFT) for VLMs? This is currently disabled in SFTrainer by reading the class of the provided processor. 

### Motivation

This would be super helpful for reducing the memory footprint of training VLMs.

### Your contribution

If everything is processed into tokens, shouldn't packing with other modalities be straightforward (ignoring memory issues with the vision encoder)? I'm not knowledgeable in how padding_free packing interacts with flash attention 2. Is this the reason it is not straightforward?

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
SFT rejected padding-free training for vision datasets. I added per-example processing and concatenation with reset positions and masked boundary labels, while rejecting models that rebuild multimodal positions. CPU tests cover image layouts and trainer wiring, and a GPU check confirmed sample isolation and finite training losses with Idefics3.
Fixes #4339.

---

> [!NOTE]
> **Medium Risk**
> Changes batching and attention boundaries for VLM training; incorrect packing could cause cross-example attention, though RoPE models are explicitly blocked and tests assert position resets and label masking.
> 
> **Overview**
> Enables **`padding_free=True`** on vision-language SFT: `SFTTrainer` wires it into `DataCollatorForVisionLanguageModeling`, which now processes each example separately, strips padding, concatenates token/image fields into one sequence, emits **`position_ids`** that restart per example (no **`attention_mask`**), and masks the first token of each segment in **`labels`**—matching the text padding-free / Flash Attention contract. Idefics-style batches get extra padding when image counts or sizes differ before concat.
> 
> The blanket ban on vision + padding-free is lifted; **`packing=True`** for vision stays unsupported. **`padding_free`** is rejected only when the model uses **multimodal RoPE** (e.g. Qwen-VL, detected via `mrope_section` in `rope_scaling`), because those models rebuild 3D positions and would break example boundaries.
> 
> Docs describe the feature, Flash Attention requirement, and RoPE limitation. Tests cover the VLM collator (Qwen2-VL, Idefics2/3), trainer init/warnings, and multimodal-RoPE rejection.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ef772ba3179107f9c65ebf2ab5b8dad03f6394d5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -1060,10 +1097,15 @@ def __init__(
             raise ValueError(
                 "Packing is not supported for vision datasets. Please set `packing=False` in the SFTConfig."
             )
-        if self._is_vision_dataset and args.padding_free:
+        if (
+            self._is_vision_dataset
+            and args.padding_free
+            and "mrope_section" in (model.config.get_text_config().rope_scaling or {})
+        ):
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/sft_trainer.py:1104]: Not reproduced. The guard does see `rope_parameters`, because `rope_scaling` is a property alias on the base config class:

```python
# transformers/configuration_utils.py:487
@property
def rope_scaling(self):
    return self.rope_parameters
```

So `model.config.get_text_config().rope_scaling` returns the same dict the rename moved the values into. I checked the predicate on two real checkpoints under transformers 5.11.0:

```
tiny-Qwen2VL    rope_scaling: {'mrope_section': [1, 1], 'rope_type': 'default', ...}   -> guard fires
tiny-Idefics2   rope_scaling: {'rope_theta': 10000.0, 'rope_type': 'default'}          -> stays silent
```

Qwen2-VL's own config keeps `mrope_section` in `rope_parameters` (`configuration_qwen2_vl.py:81` exempts it from RoPE validation), so the key survives the rename and the lookup finds it. The second half of the claim, that some text configs omit the attribute entirely, does not hold either: the property is defined on `PreTrainedConfig`, so every config that reaches this line has it.

No change here. If the alias is ever dropped upstream, the guard would need `rope_parameters` directly, but that is not the case on any version in the CI matrix.
